Sources/Data/Attributes.swift

```diff
@@ -0,0 +1,299 @@
+                default:
+                    let context = EncodingError.Context(codingPath: container.codingPath, debugDescription: "Unexpected attribute value")
+                    throw EncodingError.invalidValue(value, context)
+                }
+            }
+        }
+        var container = encoder.container(keyedBy: DynamicCodingKeys.self)
+
+        try encodeToContainer(dictionary: self.rawValue, container: &container)
+    }
+    
+    override init() {
+        rawValue = [:]
+        super.init()
+    }
+}
+
+
+typealias AttributeValue = [String: Any]
+
+extension Dictionary where Key == String, Value: Any {
+    fileprivate var asAttributesForNsDictionary: [String: Any] {
+        func transformAsNeeded(value: Any) -> Any? {
+            switch(value) {
+            case let bool as Bool:
+                // The Nextstep standard library objects do not model Bool separately from Number, so converting our attributes dictionaries to NSDictionary otherwise unmodified would lose the distinction between the two types.  We work around this by creating our own `BooleanValue` type.
+                return BooleanValue(bool)
+            case is Int, is String, is Double:
+                return value
+            default:
+                let errorMessage = "Unsupported type used in a Rover attributes dictionary: \(type(of: value))"
+                assertionFailure(errorMessage)
+                os_log("%s", type: .error, errorMessage)
+                return nil
+            }
+        }
+        
+        return self.reduce(into: [String:Any]()) { (result, element) in
+            // TODO verify element.key against "^[a-zA-Z_][a-zA-Z_0-9]*$"
+            switch(element.value) {
+            case let dictionary as Dictionary:
+                // nesting!
+                result[element.key] = dictionary.asAttributesForNsDictionary
+            case let array as Array<Any>:
+                // can only contain scalars
+                result[element.key] = array.map { transformAsNeeded(value: $0) }
+            default:
+                guard let transformed = transformAsNeeded(value: element.value) else {
+                    return
+                }
+                result[element.key] = transformed
+            }
+        }
+    }
+    
+    var attributes: NSDictionary {
+        let dictionary = NSMutableDictionary()
+        // setValuesForKeys coerces Swifty primitives to their NS* equivalents, however, in asAttributesForNsDictionary() we do a few extra transforms.
+        // setValuesForKeys' equivalent in the opposite direction is NSDictionary.dictionaryWithValues
+        let attributesSuitableForNsDictionary = self.asAttributesForNsDictionary
+        dictionary.setValuesForKeys(attributesSuitableForNsDictionary)
+        return dictionary
+    }
+    
+    /// Undo the effects of [asAttributesForNsDictionary].
+    fileprivate var fromAttributesAsNsDictionary: [String: Any] {
```

Comment:
This, along with the logic in the `NSDictionary` extension, can be moved directly into the `required init?(coder aDecoder: NSCoder)` function.